Shader-based Billboard actor

m-agour opened the pull request:
-  billboard actor using pure shader
- Implement Billboard, BillboardMaterial, and BillboardShader
- Add WGSL shder for quad generation


https://github.com/user-attachments/assets/c3eefd8d-be1b-44f9-bff7-31c26e11db0d

Example:

```python
import numpy as np
from fury import window, actor

centers = np.array([[0, 0, 0], [2, 0, 0], [-2, 0, 0]])
colors = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]])
sizes = np.array([[1.0, 1.0], [1.5, 0.8], [0.8, 1.5]])

billboard_actor = actor.billboard(
    centers=centers,
    colors=colors, 
    sizes=sizes,
    opacity=0.8
)

scene = window.Scene()
scene.add(billboard_actor)

sm = window.ShowManager(scene=scene, size=(600, 600))
sm.start()
```

Commit:
NF: added billboard actor

- Add billboard actor with camera-facing behavior using pure shader logic
- Implement Billboard world object, BillboardMaterial, and BillboardShader
- Add WGSL shader for procedural quad generation with camera orientation
- Support per-billboard positions, colors, and sizes with geometry expansion
- Include comprehensive tests for creation, rendering, and camera facing
- Isolate implementation in fury/actor/billboard.py to minimize code changes

Commit:
BF: fix colors for billboards

m-agour committed:
REF: Refactor billboard module and improve documentation

maharshi-gor reviewed (approved):
Thank you for the work @m-agour

maharshi-gor commented:
Hi @m-agour Please rebase this PR to merge this.

guaje commented on fury/tests/test_billboard.py:
```diff
@@ -0,0 +1,117 @@
+    # Test creation and geometry setup
+    centers = [[0, 0, 0], [1, 2, 3], [-2, 0.5, 4]]
+    colors = [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]
+    sizes = [[1.0, 2.0], [0.5, 0.5], [2.0, 1.0]]
```
I tried adjusting the sizes to values like [7.0, 1.0] to create rectangular billboards, but I only get square ones. Do you know if this is the intended behavior? If not, let's support square billboards in this PR and enable the rectangular form factor in a future PR.

m-agour replied:
Yes, you are right, fixed now, it was supported but while fixing the normals issue in the pipeline I disapled this.

guaje commented on fury/tests/test_billboard.py:
```diff
@@ -0,0 +1,117 @@
+    colors = [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]]
+    sizes = [[1.0, 2.0], [0.5, 0.5], [2.0, 1.0]]
+
+    bb = actor.billboard(centers, colors=colors, sizes=sizes, opacity=0.75)
+
+    # Check billboard count
+    assert hasattr(bb, "billboard_count")
+    npt.assert_equal(bb.billboard_count, 3)
+
+    # Check geometry (6 vertices per billboard)
+    geom = bb.geometry
+    npt.assert_equal(len(geom.positions.data), 18)
+    npt.assert_equal(len(geom.colors.data), 18)
+    npt.assert_equal(len(geom.texcoords.data), 18)
+    npt.assert_equal(len(geom.indices.data), 18)
```
I have added the following piece to your test:

```python
if interactive:  # pragma: no cover
        window.show([bb])
```

And I'm getting the following viz when I zoom out:

<img width="553" height="371" alt="Image" src="https://github.com/user-attachments/assets/9c622b29-236b-4dd0-ba90-0964bfe7e157" />

Which goes against these assertions.

m-agour replied:
THis is weird! where did this white IMPOSTOR come from. I tested after fixing the other things and it was gone 
<img width="2000" height="1717" alt="image" src="https://github.com/user-attachments/assets/33543f72-7545-4499-a210-92004c4eb4e7" />

guaje commented on fury/tests/test_billboard.py:
```diff
@@ -0,0 +1,117 @@
+    scene.add(bb_render)
+
+    if interactive:  # pragma: no cover
+        window.show(scene)
```
This function now expects an array of actors.

guaje commented on fury/tests/test_billboard.py:
```diff
@@ -0,0 +1,117 @@
+        window.show(scene)
+
+    # Test snapshot creation
+    tmp_file = tempfile.mktemp(suffix="_bb.png")
```
The function "mktemp" is deprecated since Python 2.3. Use `mkstemp()` or `NamedTemporaryFile(delete=False)` instead.

```python
# Test snapshot creation
tmp_file = tempfile.mkstemp(suffix="_bb.png")
arr = window.snapshot(scene=scene, fname=tmp_file[1], return_array=True)
```

guaje commented on fury/tests/test_billboard.py:
```diff
@@ -0,0 +1,117 @@
+
+    # Test snapshot creation
+    tmp_file = tempfile.mktemp(suffix="_bb.png")
+    arr = window.snapshot(scene=scene, fname=tmp_file, return_array=True)
```
This line is giving me the following error: AttributeError: 'OffscreenRenderCanvas' object has no attribute 'draw_frame'. Did you mean: '_draw_frame'?

maharshi-gor replied:
This will be fixed automatically when he rebase the PR.

guaje reviewed (commented):
Thanks for putting this together and kickstarting a new glyph system @m-agour. Here are a few issues I encountered while testing and reviewing this PR.

m-agour committed:
FIX: address PR review comments and add rectangular billboard support

m-agour commented:
Hi @maharshi-gor  @guaje, thank you for your review. I have addressed your reviews and everything is green now! except code format I rebased and still not passing even though I rebased.

maharshi-gor commented:
Hi @m-agour just run pre-commit 

```
pre-commit run --all-files
```

m-agour commented:
> Hi @m-agour just run pre-commit
> 
> ```
> pre-commit run --all-files
> ```

Did it many times already, it suggests modifing the yaml file which shouldn't be changed in this pr by me.

maharshi-gor commented:
Nope you need to update it. It is the config file for the updating the pre-commit ruff version.
@m-agour  please go ahead and push this.

m-agour committed:
BF: version mismatch fixed

m-agour commented:
done.

maharshi-gor commented:
Thanks again for the work @m-agour This opens up new opportunities for us in the FURY.
